snapcraft/internal/elf.py

```diff
@@ -282,15 +289,22 @@ def _extract(self, path: str) -> ElfDataTuple:  # noqa: C901
                     for version in versions:
                         lib.add_version(_ensure_str(version.name))
 
-            for segment in elf.iter_segments():
```

Comment:
Each segment contains data from one or more named section.  For some of the data we were extracting, it is the segment that matters rather than section name.

For example, while GNU LD writes a `.interp` section containing the dynamic loader to use, the kernel looks at the `PT_INTERP` segment to determine the loader.  An alternative compiler/linker could use a different section name and everything would work provided that the section's data was written to the `PT_INTERP` segment.

Reply:
Makes sense, thank you for the explanations!